### PR TITLE
[4.0] Move Switcher field to separate class

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

This is an alternative to https://github.com/joomla/joomla-cms/pull/26214 that aims to eliminate weird use of class attribute to set field layouts. Furthermore, Switcher has some specific logic - it should not contain more than 2 options. For this the switcher should have its own class.

### Testing Instructions

View forms containing switcher.

### Expected result

Works like before.

### Documentation Changes Required

Forms should be updated to use `type="switcher"` instead of `type="radio" class="switcher"`.
Radio field layout moved back to `layouts/joomla/form/radio.php`.

### CODE REVIEW

File: installation/forms/preinstall.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
 	<fieldset name="basic">
 		<field
 			name="ftp_enable"
-			type="radio"
+			type="Switcher"
 			label="INSTL_FTP_ENABLE_LABEL"
 			id="ftp_enable"
 			class="btn-group btn-group-reverse"
```
Comment (Quy) [inline installation/forms/preinstall.xml]: Remove?

File: plugins/editors/tinymce/form/setoptions.xml [hunk 2]
```diff
@@ -31,21 +31,19 @@
 
     <field
         name="mobile"
-        type="radio"
+        type="Switcher"
         label="PLG_TINY_FIELD_MOBILE_LABEL"
-        class="switcher"
-        default="0"
+                default="0"
```
Comment (Quy) [inline plugins/editors/tinymce/form/setoptions.xml]: Fix indentation.

File: plugins/system/debug/debug.xml [hunk 3]
```diff
@@ -259,7 +245,7 @@
 
 				<field
 					name="log_category_mode"
-					type="radio"
+					type="Switcher"
 					label="PLG_DEBUG_FIELD_LOG_CATEGORY_MODE_LABEL"
 					class="switcher switcher-danger"
```
Comment (Quy) [inline plugins/system/debug/debug.xml]:
```suggestion
```

File: installation/forms/preinstall.xml (hunk 1 above)
Comment (SharkyKZ) [reply installation/forms/preinstall.xml]: Reverted changes in this file.

**Comment by C-Lodder** [comment]
This used to be defined by `type` but was changed to a `class` as requested: https://github.com/joomla/40-backend-template/issues/171

Not sure if @brianteeman and @mbabker have the same opinion 3 years on.

**Comment by brianteeman** [comment]
I have the same opinion about it breaking b/c for no real benefit

**Comment by SharkyKZ** [comment]
Breaking B/C with what?

**Comment by dgrammatiko** [comment]
There is **NO** b/c break here!!!

@Quy why the capital letter in the type?
Everywhere else the definitions are lower case

**Comment by brianteeman** [comment]
Sorry it is three years since that thread. What I was requesting there was that it would be the class that was changed and not the type. https://github.com/joomla/40-backend-template/issues/171#issuecomment-260815802

**Comment by Quy** [comment]
Please fix conflicts.

File: administrator/components/com_csp/config.xml [hunk 4]
```diff
@@ -94,7 +89,7 @@
 				/>
 				<field
 					name="client"
-					type="radio"
+					type="switcher"
```
Comment (Quy) [inline administrator/components/com_csp/config.xml]:
```suggestion
					type="list"
```

Comment (SharkyKZ) [reply administrator/components/com_csp/config.xml]: I changed this by mistake. Reverting.

**Comment by Quy** [comment]
Enable Content - Vote plugin
Go to Configuration > Articles > List Layouts
Update the following:

Show Votes in List
Show Ratings in List

**Comment by Quy** [comment]
Create custom URL field
Update `Show URL`

**Comment by Quy** [comment]
Do new installation.
Under Database Configuration >  Connection Encryption > Two-way authentication
Update `Verify Server Certificate`

**Comment by SharkyKZ** [comment]
New fields updated.

Made a separate PR for cleaning up voting field https://github.com/joomla/joomla-cms/pull/27656.

File: libraries/src/Form/Field/SwitcherField.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+			throw new \UnexpectedValueException(sprintf('%s has no layout assigned.', $this->name));
+		}
+
+		return parent::getInput();
+	}
+
+	/**
+	 * Method to get the radio button field input markup.
+	 *
+	 * @return  string  The field input markup.
```
Comment (Fedik) [inline libraries/src/Form/Field/SwitcherField.php]: copy paste error? ;)
also for the method description

**Comment by Fedik** [comment]
it works, but have a conflict again, and liltle copy/paste mistake

File: libraries/src/Form/Field/SwitcherField.php (hunk 5 above)
Comment (SharkyKZ) [reply libraries/src/Form/Field/SwitcherField.php]: Yes, it is.

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on f414372a04a616c3573b43472e3867825a50fafc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27202">issues.joomla.org/tracker/joomla-cms/27202</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f414372a04a616c3573b43472e3867825a50fafc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27202">issues.joomla.org/tracker/joomla-cms/27202</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27202">issues.joomla.org/tracker/joomla-cms/27202</a>.</sub>

**Comment by Bakual** [comment]
I don't think this is the right thing to do.
The switcher **is** a radio field, just with a different design. It doesn't make sense to add field types for specific designs.
The current solution with multiple JLayouts is the correct solution. The technically most correct solution would be to add a "layout" attribute to the radio field declaration, like you proposed in https://github.com/joomla/joomla-cms/pull/26214. But writing `layout="joomla.form.field.radio.switcher"` into each field is tedious and not very intuitiv. Thus using a class (which is design and JS related) is a nice compromise which worked very well for the whole J3 series.

**Comment by SharkyKZ** [comment]
Using class was fine in J3. But in J4 it's used to select a layout and that's totally wrong.

>The switcher is a radio field, just with a different design. It doesn't make sense to add field types for specific designs.

Switcher does have specific logic. It must not have more than 2 options.

**Comment by Bakual** [comment]
> Using class was fine in J3. But in J4 it's used to select a layout and that's totally wrong.

It's not that much different. In J3, the layout was changed with JS, in J4 it's done in PHP.

> Switcher does have specific logic. It must not have more than 2 options.

It's not a specific logic, it's just a limitation of the same logic. Same as `btn-group-yesno`. In fact "switcher" is a direct replacement for "btn-group  btn-group-yesno".

Is it the most correct thing to do? No. But neither is this PR.
The most correct one would be the #26214, but that's horrible for developers.

**Comment by Fedik** [comment]
I think all ways are correct.

I am for a separated field. 
It gives more control, and makes thing more obvious than use of "class".
It does not hurt in the end :)

**Comment by Bakual** [comment]
> It gives more control

How that?

> It does not hurt in the end 

It's more work for extension devs to adjust their extensions to the new UI style. They can't just replace the class "btn-group-yesno" with "switcher", that part could be done with a search/replace.
They need to change the type, but not all radios - only those with 2 options (or btn-group-yesno).

So it DOES in fact hurt much.

**Comment by brianteeman** [comment]
I stick by my opinion 3 years ago as linked to further above

**Comment by Fedik** [comment]
> How that?

https://github.com/joomla/joomla-cms/blob/f414372a04a616c3573b43472e3867825a50fafc/libraries/src/Form/Field/SwitcherField.php#L91-L97

and #27672 

> It's more work for extension devs to adjust their extensions to the new UI style

No one forced to use it ASAP, it will stay a regular radio button for those for lazy.

**Comment by Bakual** [comment]
> No one forced to use it ASAP, it will stay a regular radio button for those for lazy.

The work is exactly the same, if he does it right now or someday later.
Eventually, users will expect a component to follow the core UI.

This PR is just changing for the sake of changing. There is no advantage, but only disadvantage.

**Comment by SharkyKZ** [comment]
The point is dropping a bad practice.

**Comment by Fedik** [comment]
Okay, then it should be that PR #26214 I do not mind,
but then every XML defination should have: 
`hiddenLabel="true" layout="joomla.form.field.radio.switcher"`
Is that much easier to edit than changing type to `type="switcher"`?

But we should not use a class to switch the layout, it bad example for other.

I think this PR is good.

**Comment by Bakual** [comment]
> The point is dropping a bad practice.

It's not bad practice. It's a good compromise because the technically correct solutions are worse.

**Comment by rdeutz** [comment]
tbh I don't like the change, a switcher is just a radio with two options, why giving anything a specific name, just my optionen.

@wilsonge your turn

**Comment by rdeutz** [comment]
needs review by @wilsonge <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27202">issues.joomla.org/tracker/joomla-cms/27202</a>.</sub>

**Comment by Quy** [comment]
@SharkyKZ OK to close now that #26214 has been merged?

**Comment by SharkyKZ** [comment]
No, this is still the best option.

**Comment by Bakual** [comment]
It's not, the other one is the technical correct solution, allthought ugly. This one would be worse.

**Comment by wilsonge** [comment]
Personally I like the layout attribute - I think it's a good middle ground - this is still radio buttons at the end of the day (in html too) - we're just adding some crazy css styling to them. Yes it's specific styling for when you have only two options - but how many options and the layout are both decided in the XML already anyhow

**Comment by wilsonge** [comment]
Choosing to reject this. I still prefer the layouts option as the middle group after a bit of time to see it in action

**Pull request closed without merging**